libmtd/mtd.c

```diff
@@ -0,0 +1,466 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Memory Technology Device (MTD) Interface
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "include/mtd/mtd.h"
+
+#include <errno.h>
+#include <stdlib.h>
+
+
+int mtd_erase(struct mtd_info *mtdInfo, struct erase_info *instr)
+{
+	int res;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (mtd->ops->erase == NULL) {
+		return -EOPNOTSUPP;
+	}
+
+	if (instr->addr >= mtdInfo->size || instr->len > (mtdInfo->size - instr->addr)) {
+		return -EINVAL;
+	}
+
+	if ((mtdInfo->flags & MTD_WRITEABLE) == 0) {
+		return -EROFS;
+	}
+
+	instr->fail_addr = MTD_FAIL_ADDR_UNKNOWN;
+	instr->state = MTD_ERASING;
+	if (instr->len == 0) {
+		instr->state = MTD_ERASE_DONE;
+		if (instr->callback != NULL) {
+			instr->callback(instr);
+		}
+
+		return EOK;
+	}
+
+	instr->state = MTD_ERASE_PENDING;
+	res = mtd->ops->erase(mtdInfo->storage, instr->addr, instr->len);
+	if (res < 0) {
+		instr->fail_addr = instr->addr;
+		instr->state = MTD_ERASE_FAILED;
+	}
+	else {
+		instr->state = MTD_ERASE_DONE;
+	}
+
+	if (instr->callback != NULL) {
+		instr->callback(instr);
+	}
+
+	return EOK;
+}
+
+
+int mtd_point(struct mtd_info *mtdInfo, off_t from, size_t len, size_t *retlen, void **virt, size_t *phys)
+{
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	*retlen = 0;
+	*virt = NULL;
+
+	if (phys != NULL) {
+		*phys = 0;
+	}
+
+	if (mtd->ops->point == NULL) {
+		return -EOPNOTSUPP;
+	}
+
+	if (from < 0 || from >= mtdInfo->size || len > (mtdInfo->size - from)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return EOK;
+	}
+
+	return mtd->ops->point(mtdInfo->storage, from, len, retlen, virt, phys);
```

Comment:
Avoiding multiple exit paths
```suggestion
	if (mtd->ops->point == NULL) {
		ret = -EOPNOTSUPP;
	}
	else if (from < 0 || from >= mtdInfo->size || len > (mtdInfo->size - from)) {
		ret = -EINVAL;
	}
	else if (len == 0) {
		ret = EOK;
	}
	else {
		ret = mtd->ops->point(mtdInfo->storage, from, len, retlen, virt, phys);
	}
	
	return ret;
```

Reply:
Most other cases can be rearranged in a similar fasion